kartothek/io_components/metapartition.py

```diff
@@ -1233,10 +1233,20 @@ def build_indices(self, columns):
                     )
                 )
 
+            # There is at least one table with this column (see check above), so we can get the dtype from there. Also,
+            # shared dtypes are ensured to be compatible.
+            dtype = list(
+                meta.field(col).type
+                for meta in self.table_meta.values()
+                if col in meta.names
+            )[0]
+
             new_index = ExplicitSecondaryIndex(
-                column=col, index_dct={value: [self.label] for value in possible_values}
+                column=col,
+                index_dct={value: [self.label] for value in possible_values},
+                dtype=dtype,
```

Comment:
We should've seen the lack of dtype as a warning somewhere, I believe. Did you evaluate whether this pops up somewhere else?

Reply:
So normally this is OK, the index will infer the dtype during the store operation from `index_dct` (this even works for ints under python 3, since numpy ints have their own type now) or during load (from the parquet file).

But there is one edge case during index build where this does not work (see last commit): when the `index_dct` is empty. In that case, the index constructor even raises, since the type inference would obviously be bollocks. I am somewhat surprised that nobody ran into this ("create index where one partition has an empty DF").

Also, who emits the warning?

Reply:
Sorry, I was mistaken. I thought we disallowed this / deprecated this but that's not the case.